File: Project 4/python/simpletron.py
#!/usr/bin/python3
import sys

memory = [0] * 100
accumulator = 0
programCounter = 0

def printStatus():
    print("\nAccumulator: ", accumulator)
    print("programCounter: ", str(programCounter).zfill(2))
    print("memory: ", memory)

def printHelp():
    print("to execute from file: python3 simpletron.py [file]")
    print("enter \"exit\" to quit")
    print("enter \"input\" to enter instruction codes")
    print("enter \"end\" to save instruction codes into memory")
    print("enter \"run\" to execute instructions saved in memory\n")

def saveline(line):
    memory[programCounter] = line

def execute_line(instruction):
    global accumulator
    global programCounter
    branch = False
    opCode = instruction // 100
    operand = instruction % 100
    if opCode == 40:
        programCounter = operand
        branch = True
    elif opCode == 41:
        if accumulator < 0:
            programCounter = operand
            branch = True
    elif opCode == 42:
        if accumulator == 0:
            programCounter = operand
            branch = True
    elif opCode == 43:
        print("Execution Halted! Exiting...")
        sys.exit()
    elif opCode == 30:
        accumulator = accumulator + memory[operand]
    elif opCode == 31:
        accumulator = accumulator - memory[operand]
    elif opCode == 32:
        accumulator = accumulator * memory[operand]
    elif opCode == 33:
        accumulator = accumulator / memory[operand]
    elif opCode == 10:
        numInput = input("Read number: ")
        try:
            read = int(numInput)
            memory[operand] = read
        except ValueError:
            print("ERROR please try again")
    elif opCode == 11:
        print(memory[operand])
    elif opCode == 20:
        accumulator = memory[operand]
    elif opCode == 21:
        memory[operand] = accumulator

    if branch:
        execute_line(memory[programCounter])

def execute_instructions():
    global memory
    global programCounter
    programCounter = 0
    while memory[programCounter] != 0:
        print("executing {:02d}# ".format(programCounter), memory[programCounter])
        execute_line(memory[programCounter])
        programCounter = programCounter + 1

def interactive():
    global programCounter
    isReadingInstructions = False
    isInstructionSaved = False
    print("\nEntering Simpletron in interactive mode...\n")
    printHelp()
    while 1:
        userIn = input("{:02d}# ".format(programCounter) if isReadingInstructions else ">>>")
        try:
            userInput = int(userIn)
            if isReadingInstructions and not isInstructionSaved:
                saveline(userInput)
                programCounter = programCounter + 1
        except ValueError:
            if userIn == "exit":
                sys.exit()
            elif userIn == "status":
                printStatus()
            elif userIn == "input":
                isReadingInstructions = True
            elif userIn == "end":
                isInstructionSaved = True
                isReadingInstructions = False
            elif userIn == "run":
                execute_instructions()
            else:
                printHelp()

def load_script(filename):
    global memory
    global programCounter
    lineNumber = 0
    sml_script = open(filename, 'r')

    for line in sml_script:
        line = line.strip()
        memory[lineNumber] = int(line)
        lineNumber = lineNumber + 1

    execute_instructions()

def main():
    arr = sys.argv[1:]
    if len(arr) == 0:
        interactive()
    else:
        load_script(arr[0])

if __name__ == "__main__" : main()

File: Project 4/java/Simpletron.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;

class Simpletron {

	private static int ACC = 0;
	private static int[] MEMORY = new int[100];
	private static int programCounter = 0;
	private static boolean readInstructions = true;

	private static void interactive () {
		System.out.println("Enter instruction codes. when finished, input a negative number");
		int inputCode = 0;
		BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
		while (readInstructions) {
			try {
				System.out.print("::#");
				String input = reader.readLine();
				inputCode = Integer.parseInt(input);
				
				if (inputCode < 0) {
					return;
				} else if (inputCode < 1000 || inputCode > 4300) {
					System.out.println("try again...");
				} else {
					MEMORY[programCounter] = inputCode;
					programCounter++;
				}
			} catch (Exception e) {
				System.out.println("input error: " + e);
			}
		}
	}
	
	static void printInput () {
		for (int i = 0; i < MEMORY.length && MEMORY[i] != 0; i++) {
			System.out.print(MEMORY[i] + " ");
		}
		System.out.print("\n");
	}
	
	static String askUserInput (String text) {
		System.out.print(text);
		BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
		String input = "weird";
		try {
			input = reader.readLine();
			return input;
		} catch (Exception e) {
			System.out.println("input error: " + e);
		}
		return input;
	}
	
	static int convertStringToInt (String input) {
		int returnInt = 0;
		try {
			returnInt = Integer.parseInt(input);
		} catch (Exception e) {
			System.out.println("input error: " + e);
		}
		return returnInt;
	}
	
	static void executeFromMemory () {
		programCounter = 0;
		while (MEMORY[programCounter] !=0) {
			executeLine(MEMORY[programCounter]);
			programCounter++;
		}
	}
	
	static void executeLine (int code) {
		int opCode = code / 100;
		int operand = code % 100;
		switch (opCode) {
			case 40:
				programCounter = operand;
				executeLine(MEMORY[programCounter]);
				break;
			case 41:
				if (ACC < 0) {
					programCounter = operand;
					executeLine(MEMORY[programCounter]);
				}
				break;
			case 42:
				if (ACC == 0) {
					programCounter = operand;
					executeLine(MEMORY[programCounter]);
				}
				break;
			case 43:
				System.exit(0);
				break;
			case 30:
				ACC += MEMORY[operand];
				break;
			case 31:
				ACC -= MEMORY[operand];
				break;
			case 32:
				ACC *= MEMORY[operand];
				break;
			case 33:
				ACC /= MEMORY[operand];
				break;
			case 10:
				String read = askUserInput("Read: ");
				int readCode = convertStringToInt(read);
				MEMORY[operand] = readCode;
				break;
			case 11:
				System.out.println(MEMORY[operand]);
				break;
			case 20:
				ACC = MEMORY[operand];
				break;
			case 21:
				MEMORY[operand] = ACC;
				break;
		}
	}
	
	private static void openSMLFile(String filename) {
		try {
			File f=new File(filename);     
			FileReader fr=new FileReader(f);   
			BufferedReader br=new BufferedReader(fr);  
			String s=null;  
			int i = 0;        
			while ((s=br.readLine())!=null) {
				System.out.println(s);
				MEMORY[i] = convertStringToInt(s);
				i++;
			}
			fr.close();
		} catch(Exception e) {
			System.out.println("file open error: " + e);
		}
	}
	
	public static void main (String[] args) {
		if (args.length == 0) {
			System.out.println("No file entered. Entering interactive mode...");
			interactive();
			//printInput();
			System.out.println("Executing...");
			executeFromMemory();
			
		} else {
			openSMLFile(args[0]);
			printInput();
			executeFromMemory();
		}
		
	}
}
